neo/SmartContract/Native/Tokens/Nep5Token.cs

```diff
@@ -73,6 +73,32 @@ internal protected virtual void MintTokens(ApplicationEngine engine, UInt160 acc
             engine.SendNotification(ScriptHash, new StackItem[] { "Transfer", StackItem.Null, account.ToArray(), amount });
         }
 
+        internal protected virtual void Burn(ApplicationEngine engine, UInt160 account, BigInteger amount)
+        {
+            if (amount.Sign < 0) throw new ArgumentOutOfRangeException(nameof(amount));
+            if (amount.IsZero) return;
+            StorageKey key = CreateAccountKey(account);
+            StorageItem storage = engine.Snapshot.Storages.GetAndChange(key);
+            TState state = new TState();
+            state.FromByteArray(storage.Value);
+            if (state.Balance < amount) throw new InvalidOperationException();
+            OnBalanceChanging(engine, account, state, -amount);
+            if (state.Balance == amount)
+            {
+                engine.Snapshot.Storages.Delete(key);
+            }
+            else
+            {
+                state.Balance -= amount;
+                storage.Value = state.ToByteArray();
+            }
+            storage = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_TotalSupply));
+            BigInteger totalSupply = new BigInteger(storage.Value);
+            totalSupply -= amount;
+            storage.Value = totalSupply.ToByteArray();
+            engine.SendNotification(ScriptHash, new StackItem[] { "Transfer", account.ToArray(), StackItem.Null, amount });
```

Comment:
Maybe we can use specific notification for mint and burn, no reuse the Transfer

Reply:
https://github.com/neo-project/proposals/blob/master/nep-5.mediawiki#transfer-1

> A token contract which creates new tokens MUST trigger a transfer event with the from address set to null when tokens are created.

>A token contract which burns tokens MUST trigger a transfer event with the to address set to null when tokens are burned.